kafka/producer.go

```diff
@@ -17,4 +17,164 @@
+type ProducerConfig struct {
+	// Broker holds the (host:port) address of the Kafka broker to which
+	// events should be published.
+	Broker string
+
+	// ClientID to use when connecting to Kafka. This is used for logging
+	// and client identification purposes.
+	ClientID string
+	// Version is the software version to use in the Kafka client. This is
+	// useful since it shows up in Kafka metrics and logs.
+	Version string
+
+	// Logger is used for logging producer errors.
+	Logger *zap.Logger
+
+	// Sync can be used to indicate whether production should be synchronous.
+	Sync bool
+
+	Mutators []RecordMutator
+}
+
+// Validate checks that cfg is valid, and returns an error otherwise.
+func (cfg ProducerConfig) Validate() error {
+	if cfg.Broker == "" {
+		return errors.New("broker cannot be empty")
+	}
+	if cfg.Logger == nil {
+		return errors.New("logger cannot be nil")
+	}
+	return nil
+}
+
+// Producer is a model.BatchProcessor that publishes events to Kafka.
+type Producer struct {
+	cfg    ProducerConfig
+	client *kgo.Client
+
+	mu sync.RWMutex
+}
+
+// NewProducer returns a new Producer with the given config.
+func NewProducer(cfg ProducerConfig) (*Producer, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid producer config: %w", err)
+	}
+
+	opts := []kgo.Opt{
+		kgo.SeedBrokers(cfg.Broker),
+		kgo.WithLogger(kzap.New(cfg.Logger)),
+	}
+	if cfg.ClientID != "" {
+		opts = append(opts, kgo.ClientID(cfg.ClientID))
+		if cfg.Version != "" {
+			opts = append(opts, kgo.SoftwareNameAndVersion(
+				cfg.ClientID, cfg.Version,
+			))
+		}
+	}
+	// TODO(marclop) block on re-balances, auto-commit high watermarks.
+	client, err := kgo.NewClient(opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating producer: %w", err)
+	}
+	// Issue a metadata refresh request on construction, so the broker list is
+	// populated.
+	client.ForceMetadataRefresh()
+
+	return &Producer{
+		cfg:    cfg,
+		client: client,
+	}, nil
+}
+
+// Close stops the producer
+func (p *Producer) Close() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.client.Close()
+	return nil
+}
+
+// ProcessBatch publishes the events in batch to the specified Kafka topic.
+func (p *Producer) ProcessBatch(ctx context.Context, batch *model.Batch) error {
+	// Take a read lock to prevent Close from closing the input channel
+	// while we're attempting to send to it.
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	var headers []kgo.RecordHeader
+	if m, ok := queuecontext.MetadataFromContext(ctx); ok {
+		for k, v := range m {
+			headers = append(headers, kgo.RecordHeader{
+				Key:   k,
+				Value: []byte(v),
+			})
+		}
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(*batch))
+
+	for _, event := range *batch {
+		record := &kgo.Record{}
+
+		if len(headers) != 0 {
+			record.Headers = headers
+		}
+
+		for _, rm := range p.cfg.Mutators {
+			rm(event, record)
+		}
+
+		encoded, err := event.MarshalJSON()
```

Comment:
Can you introduce an `Encoder` interface in this package and use the `codec.JSON` as the producer's encoder?

```go
// Encoder encodes a model.APMEvent to a []byte
type Encoder interface {
	// Encode accepts a model.APMEvent and returns the encoded representation.
	Encode(model.APMEvent) ([]byte, error)
}
```

Then this would call the decoder instead of making assumptions on the encoding:
```suggestion
		encoded, err := p.cfg.Encoder.Encode(event)
```

Reply:
Thanks for this! I missed the codec changes :facepalm: